File: src/revl/lsp/protocol.py
"""The LSP wire protocol: `Content-Length` framed JSON-RPC 2.0 over a byte
stream.

Hand-rolled, with no third-party dependency — the same stance the MCP server
takes. The framing is the only part of the language server that touches raw
bytes, so it is isolated here and driven over `BytesIO` in tests; the server
above it deals only in decoded messages.

`read_message` separates the two ways a read can fail to produce a message
(issue #314). EOF returns None and the loop stops. A frame that arrived but
could not be decoded raises `MalformedFrame`: the peer is still connected, so
it is owed a `-32700` and the loop must keep reading. Collapsing the two — as
this module used to — turned one bad byte into a silent exit 0, which an editor
sees as the language server dying.
"""

from __future__ import annotations

import json

_HEADER_SEP = b"\r\n"
_CONTENT_LENGTH = b"Content-Length:"


class MalformedFrame(Exception):
    """A frame arrived but could not be decoded into a message.

    Raised instead of returning None so the caller can tell this apart from
    EOF: the peer is still there and is owed a `-32700` parse error, and the
    read loop has to continue rather than exit.
    """


def read_message(stream) -> dict | None:
    """Read one framed JSON-RPC message from a binary stream.

    Returns the decoded message, or None at EOF — including a frame truncated
    by EOF, where there is no longer a peer to answer.

    Raises `MalformedFrame` when a complete-looking frame cannot be decoded: a
    non-integer or negative `Content-Length`, headers with no `Content-Length`
    at all, or a body that is not valid UTF-8 JSON. Only the last of those
    leaves the stream on a frame boundary; the others resynchronize by reading
    the remaining bytes as headers, which terminates because every iteration
    consumes at least one line and EOF ends the loop.
    """
    length: int | None = None
    while True:
        line = stream.readline()
        if not line:
            return None  # EOF, mid-headers or between frames
        if line in (_HEADER_SEP, b"\n"):
            break  # end of headers
        if line.lower().startswith(_CONTENT_LENGTH.lower()):
            raw = line.split(b":", 1)[1].strip()
            try:
                length = int(raw)
            except ValueError:
                raise MalformedFrame(
                    f"Content-Length is not an integer: {raw!r}") from None
    if length is None:
        raise MalformedFrame("frame carries no Content-Length header")
    if length < 0:
        # `stream.read(-1)` would swallow the rest of the stream and then try
        # to parse it as one body.
        raise MalformedFrame(f"Content-Length is negative: {length}")
    body = stream.read(length)
    if body is None or len(body) < length:
        return None  # truncated at EOF
    try:
        return json.loads(body.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise MalformedFrame(f"body is not valid UTF-8 JSON: {exc}") from None


def write_message(stream, message: dict) -> None:
    """Serialize and frame one message onto a binary stream."""
    body = json.dumps(message).encode("utf-8")
    header = f"Content-Length: {len(body)}\r\n\r\n".encode("ascii")
    stream.write(header)
    stream.write(body)
    stream.flush()


def response(request_id, result) -> dict:
    return {"jsonrpc": "2.0", "id": request_id, "result": result}


def error(request_id, code: int, message: str) -> dict:
    return {"jsonrpc": "2.0", "id": request_id, "error": {"code": code, "message": message}}


def notification(method: str, params: dict) -> dict:
    return {"jsonrpc": "2.0", "method": method, "params": params}

File: crates/revl-lsp/src/protocol.rs
//! `Content-Length` framed JSON-RPC 2.0 over a byte stream.
//!
//! The rust twin of `src/revl/lsp/protocol.py`, hand-rolled for the same reason
//! the reference hand-rolls it: the framing is the only part of the server that
//! touches raw bytes, and item 336's exit test is byte-identity with
//! `python -m revl.lsp`, which a protocol crate's own serializer would not give
//! (it re-encodes messages with its own separators and key order). Reading
//! mirrors the reference's tolerances exactly, including returning `None` on a
//! truncated or malformed frame so the loop stops cleanly.

use std::io::{self, BufRead, Write};

use serde_json::{json, Value};

use crate::pyjson;

const CONTENT_LENGTH: &str = "content-length:";

/// Read one framed message, or `None` at EOF / on a malformed frame.
pub fn read_message<R: BufRead>(stream: &mut R) -> Option<Value> {
    let mut length: Option<usize> = None;
    loop {
        let mut line = Vec::new();
        let read = stream.read_until(b'\n', &mut line).ok()?;
        if read == 0 {
            return None;
        }
        if line == b"\r\n" || line == b"\n" {
            break; // end of headers
        }
        let text = String::from_utf8_lossy(&line);
        if text.to_ascii_lowercase().starts_with(CONTENT_LENGTH) {
            length = text
                .split_once(':')
                .and_then(|(_, value)| value.trim().parse::<usize>().ok());
        }
    }
    let length = length?;
    let mut body = vec![0u8; length];
    read_exact_or_none(stream, &mut body)?;
    serde_json::from_slice(&body).ok()
}

fn read_exact_or_none<R: BufRead>(stream: &mut R, buffer: &mut [u8]) -> Option<()> {
    let mut filled = 0;
    while filled < buffer.len() {
        match io::Read::read(stream, &mut buffer[filled..]) {
            Ok(0) => return None, // truncated frame at EOF
            Ok(n) => filled += n,
            Err(ref err) if err.kind() == io::ErrorKind::Interrupted => {}
            Err(_) => return None,
        }
    }
    Some(())
}

/// Serialize and frame one message, in the reference's exact bytes.
pub fn write_message<W: Write>(stream: &mut W, message: &Value) -> io::Result<()> {
    let body = pyjson::dumps(message);
    write!(stream, "Content-Length: {}\r\n\r\n", body.len())?;
    stream.write_all(&body)?;
    stream.flush()
}

pub fn response(request_id: &Value, result: Value) -> Value {
    json!({"jsonrpc": "2.0", "id": request_id, "result": result})
}

pub fn error(request_id: &Value, code: i64, message: &str) -> Value {
    json!({
        "jsonrpc": "2.0",
        "id": request_id,
        "error": {"code": code, "message": message},
    })
}

pub fn notification(method: &str, params: Value) -> Value {
    json!({"jsonrpc": "2.0", "method": method, "params": params})
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    #[test]
    fn a_framed_message_round_trips() {
        let message = json!({"jsonrpc": "2.0", "id": 1, "method": "initialize"});
        let mut buffer = Vec::new();
        write_message(&mut buffer, &message).unwrap();
        assert!(buffer.starts_with(b"Content-Length: "));
        let mut cursor = Cursor::new(buffer);
        assert_eq!(read_message(&mut cursor).unwrap(), message);
        assert!(read_message(&mut cursor).is_none()); // clean EOF
    }

    #[test]
    fn a_truncated_body_stops_the_loop() {
        let mut cursor = Cursor::new(b"Content-Length: 40\r\n\r\n{\"a\":1}".to_vec());
        assert!(read_message(&mut cursor).is_none());
    }

    #[test]
    fn the_frame_uses_python_separators() {
        let mut buffer = Vec::new();
        write_message(&mut buffer, &json!({"a": 1, "b": 2})).unwrap();
        let text = String::from_utf8(buffer).unwrap();
        assert!(text.ends_with("{\"a\": 1, \"b\": 2}"), "{text}");
        assert!(text.starts_with("Content-Length: 16\r\n\r\n"), "{text}");
    }
}
